File: problems/hackerrank/arrays/3DSurfaceArea/3DSurfaceArea.py
#!/bin/python3

def surfaceArea(A):
    """
    Calculates the 3D surface area of a toy geometry described by a 2D grid.
    
    Args:
        A (list of list of int): A 2D array representing the height of cubes in each cell.
        
    Returns:
        int: The total 3D surface area of the shape.
    """
    H = len(A)
    W = len(A[0])

    area = 2 * H * W

    for i in range(H):
        for j in range(W):
            h = A[i][j]

            up = A[i-1][j] if i > 0 else 0
            area += max(h - up, 0)

            down = A[i+1][j] if i < H-1 else 0
            area += max(h - down, 0)

            left = A[i][j-1] if j > 0 else 0
            area += max(h - left, 0)

            right = A[i][j+1] if j < W-1 else 0
            area += max(h - right, 0)

    return area

if __name__ == '__main__':
    H, W = map(int, input().split())

    A = []
    for _ in range(H):
        A.append(list(map(int, input().split())))

    result = surfaceArea(A)
    print(result)

File: problems/hackerrank/arrays/3DSurfaceArea/3DSurfaceArea.go
package main

import (
	"fmt"
)

// surfaceArea calculates the 3D surface area of a toy geometry described by a 2D grid.
// It takes a 2D slice of integers representing the height of cubes in each cell,
// and returns the total 3D surface area of the shape.
func surfaceArea(A [][]int) int {
	H := len(A)
	if H == 0 {
		return 0
	}
	W := len(A[0])

	area := 2 * H * W

	for i := 0; i < H; i++ {
		for j := 0; j < W; j++ {
			h := A[i][j]

			// Up
			up := 0
			if i > 0 {
				up = A[i-1][j]
			}
			if h > up {
				area += h - up
			}

			// Down
			down := 0
			if i < H-1 {
				down = A[i+1][j]
			}
			if h > down {
				area += h - down
			}

			// Left
			left := 0
			if j > 0 {
				left = A[i][j-1]
			}
			if h > left {
				area += h - left
			}

			// Right
			right := 0
			if j < W-1 {
				right = A[i][j+1]
			}
			if h > right {
				area += h - right
			}
		}
	}

	return area
}

func main() {
	var H, W int
	_, err := fmt.Scan(&H, &W)
	if err != nil {
		return
	}

	A := make([][]int, H)
	for i := 0; i < H; i++ {
		A[i] = make([]int, W)
		for j := 0; j < W; j++ {
			fmt.Scan(&A[i][j])
		}
	}

	result := surfaceArea(A)
	fmt.Println(result)
}
